ponyc .deb package gives ld.gold link error on ubuntu 16.10, recommends recompile with -fPIC. --pic don't work with .deb ponyc

On ubuntu 16.10 (updated on 23 oktobober 2016) the .deb package for ponyc gives the following error:

ponyc -v
0.7.0-4986123 [release] (llvm 3.9.0) (gcc-5 (Ubuntu 5.4.1-2ubuntu2) 5.4.1 20160929)

peter@peter-UX305CA:~/Dokument/programs/pony/ponyc$ ponyc examples/helloworld/
Building builtin -> /usr/lib/pony/0.7.0-2189.4986123/packages/builtin
Building examples/helloworld/ -> /home/peter/Dokument/programs/pony/ponyc/examples/helloworld
Generating
 Reachability
 Selector painting
 Data prototypes
 Data types
 Function prototypes
 Functions
 Descriptors
Optimising
Writing ./helloworld.o
Linking ./helloworld
Warning: environment variable $CC undefined, using gcc-5 as the linker
/usr/bin/ld.gold: error: ./helloworld.o: requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: ./helloworld.o: requires dynamic R_X86_64_32 reloc against 'pony_personality_v0' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(start.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(scheduler.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(cpu.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(serialise.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(cycle.o): requires dynamic R_X86_64_32 reloc against 'ponyint_pool_free_size' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(trace.o): requires unsupported dynamic reloc 11; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(delta.o): requires dynamic R_X86_64_32 reloc against 'ponyint_pool_alloc_size' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(actormap.o): requires dynamic R_X86_64_32 reloc against 'ponyint_pool_alloc_size' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(options.o): requires dynamic R_X86_64_PC32 reloc against 'strncmp' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(heap.o): requires unsupported dynamic reloc 11; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(pool.o): requires unsupported dynamic reloc 11; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(alloc.o): requires dynamic R_X86_64_PC32 reloc against 'mmap' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(pagemap.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(posix_except.o): requires unsupported dynamic reloc 11; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(socket.o): requires dynamic R_X86_64_PC32 reloc against 'getaddrinfo' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(stdfd.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(lsda.o): requires dynamic R_X86_64_PC32 reloc against '_Unwind_GetLanguageSpecificData' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(threads.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(epoll.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(asio.o): requires dynamic R_X86_64_32 reloc against 'ponyint_asio_backend_dispatch' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(actor.o): requires dynamic R_X86_64_PC32 reloc against 'memset' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(fun.o): requires unsupported dynamic reloc 11; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(hash.o): requires dynamic R_X86_64_PC32 reloc against 'memset' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(objectmap.o): requires dynamic R_X86_64_32 reloc against 'ponyint_pool_alloc_size' which may overflow at runtime; recompile with -fPIC
collect2: error: ld returned 1 exit status
Error:
unable to link: gcc-5 -o ./helloworld -O3 -march=native -mcx16 -fuse-ld=gold  ./helloworld.o -L"/usr/lib/pony/0.7.0-2189.4986123/bin/" -Wl,-rpath,"/usr/lib/pony/0.7.0-2189.4986123/bin/" -L"/usr/lib/pony/0.7.0-2189.4986123/bin/../lib" -Wl,-rpath,"/usr/lib/pony/0.7.0-2189.4986123/bin/../lib" -L"/usr/lib/pony/0.7.0-2189.4986123/bin/../packages" -Wl,-rpath,"/usr/lib/pony/0.7.0-2189.4986123/bin/../packages" -L"/usr/local/lib" -Wl,-rpath,"/usr/local/lib" -Wl,--start-group -Wl,--end-group  -lponyrt -lpthread -ldl   -lm

************************************************************************************************

Compiling with --pic gives:

peter@peter-UX305CA:~/Dokument/programs/pony/ponyc$ ponyc examples/helloworld/ --pic
Building builtin -> /usr/lib/pony/0.7.0-2189.4986123/packages/builtin
Building examples/helloworld/ -> /home/peter/Dokument/programs/pony/ponyc/examples/helloworld
Generating
 Reachability
 Selector painting
 Data prototypes
 Data types
 Function prototypes
 Functions
 Descriptors
Optimising
Writing ./helloworld.o
Linking ./helloworld
Warning: environment variable $CC undefined, using gcc-5 as the linker
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(start.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(scheduler.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(cpu.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(serialise.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(cycle.o): requires dynamic R_X86_64_32 reloc against 'ponyint_pool_free_size' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(trace.o): requires unsupported dynamic reloc 11; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(delta.o): requires dynamic R_X86_64_32 reloc against 'ponyint_pool_alloc_size' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(actormap.o): requires dynamic R_X86_64_32 reloc against 'ponyint_pool_alloc_size' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(options.o): requires dynamic R_X86_64_PC32 reloc against 'strncmp' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(heap.o): requires unsupported dynamic reloc 11; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(pool.o): requires unsupported dynamic reloc 11; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(alloc.o): requires dynamic R_X86_64_PC32 reloc against 'mmap' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(pagemap.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(posix_except.o): requires unsupported dynamic reloc 11; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(socket.o): requires dynamic R_X86_64_PC32 reloc against 'getaddrinfo' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(stdfd.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(lsda.o): requires dynamic R_X86_64_PC32 reloc against '_Unwind_GetLanguageSpecificData' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(threads.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(epoll.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(asio.o): requires dynamic R_X86_64_32 reloc against 'ponyint_asio_backend_dispatch' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(actor.o): requires dynamic R_X86_64_PC32 reloc against 'memset' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(fun.o): requires unsupported dynamic reloc 11; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(hash.o): requires dynamic R_X86_64_PC32 reloc against 'memset' which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: /usr/lib/pony/0.7.0-2189.4986123/bin/../lib/libponyrt.a(objectmap.o): requires dynamic R_X86_64_32 reloc against 'ponyint_pool_alloc_size' which may overflow at runtime; recompile with -fPIC
collect2: error: ld returned 1 exit status
Error:
unable to link: gcc-5 -o ./helloworld -O3 -march=native -mcx16 -fuse-ld=gold  ./helloworld.o -L"/usr/lib/pony/0.7.0-2189.4986123/bin/" -Wl,-rpath,"/usr/lib/pony/0.7.0-2189.4986123/bin/" -L"/usr/lib/pony/0.7.0-2189.4986123/bin/../lib" -Wl,-rpath,"/usr/lib/pony/0.7.0-2189.4986123/bin/../lib" -L"/usr/lib/pony/0.7.0-2189.4986123/bin/../packages" -Wl,-rpath,"/usr/lib/pony/0.7.0-2189.4986123/bin/../packages" -L"/usr/local/lib" -Wl,-rpath,"/usr/local/lib" -Wl,--start-group -Wl,--end-group  -lponyrt -lpthread -ldl   -lm

***********************************************************************************************'
ponyc compiled from source:
$ ./build/release/ponyc -v
0.7.0-e225e47 [release] (llvm 3.8.1) (cc (Ubuntu 6.2.0-5ubuntu12) 6.2.0 20161005)

peter@peter-UX305CA:~/Dokument/programs/pony/ponyc$ ./build/release/ponyc examples/helloworld/ 
Building builtin -> /home/peter/Dokument/programs/pony/ponyc/packages/builtin
Building examples/helloworld/ -> /home/peter/Dokument/programs/pony/ponyc/examples/helloworld
Generating
 Reachability
 Selector painting
 Data prototypes
 Data types
 Function prototypes
 Functions
 Descriptors
Optimising
Writing ./helloworld.o
Linking ./helloworld
Warning: environment variable $CC undefined, using cc as the linker
/usr/bin/ld.gold: error: ./helloworld.o: requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/bin/ld.gold: error: ./helloworld.o: requires dynamic R_X86_64_32 reloc against 'pony_personality_v0' which may overflow at runtime; recompile with -fPIC
collect2: error: ld returned 1 exit status
Error:
unable to link: cc -o ./helloworld -O3 -march=native -mcx16 -fuse-ld=gold  ./helloworld.o -L"/home/peter/Dokument/programs/pony/ponyc/build/release/" -Wl,-rpath,"/home/peter/Dokument/programs/pony/ponyc/build/release/" -L"/home/peter/Dokument/programs/pony/ponyc/build/release/../../packages" -Wl,-rpath,"/home/peter/Dokument/programs/pony/ponyc/build/release/../../packages" -L"/usr/local/lib" -Wl,-rpath,"/usr/local/lib" -Wl,--start-group -Wl,--end-group  -lponyrt -lpthread -ldl   -lm

*************************************************************************************

Building with --pic works for ponyc built from source:

peter@peter-UX305CA:~/Dokument/programs/pony/ponyc$ ./build/release/ponyc examples/helloworld/ --pic
Building builtin -> /home/peter/Dokument/programs/pony/ponyc/packages/builtin
Building examples/helloworld/ -> /home/peter/Dokument/programs/pony/ponyc/examples/helloworld
Generating
 Reachability
 Selector painting
 Data prototypes
 Data types
 Function prototypes
 Functions
 Descriptors
Optimising
Writing ./helloworld.o
Linking ./helloworld
Warning: environment variable $CC undefined, using cc as the linker
$

peter@peter-UX305CA:~/Dokument/programs/pony/ponyc$ cc -v
Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/6/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 6.2.0-5ubuntu12' --with-bugurl=file:///usr/share/doc/gcc-6/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-6 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-6-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-6-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-6-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 6.2.0 20161005 (Ubuntu 6.2.0-5ubuntu12)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,8 @@ All notable changes to the Pony compiler and standard library will be documented
 
 ### Fixed
 
+- Link the correct version of `libponyrt` when compiling with `--pic` on Linux (issue #1359)
+
 ### Added
 
 ### Changed

--- a/Makefile
+++ b/Makefile
@@ -580,6 +580,9 @@ install: libponyc libponyrt ponyc
 	@mkdir -p $(destdir)/lib
 	@mkdir -p $(destdir)/include/pony/detail
 	$(SILENT)cp $(PONY_BUILD_DIR)/libponyrt.a $(destdir)/lib
+ifeq ($(OSTYPE),linux)
+	$(SILENT)cp $(PONY_BUILD_DIR)/libponyrt-pic.a $(destdir)/lib
+endif
 ifneq ($(wildcard $(PONY_BUILD_DIR)/libponyrt.bc),)
 	$(SILENT)cp $(PONY_BUILD_DIR)/libponyrt.bc $(destdir)/lib
 endif
@@ -594,6 +597,9 @@ ifeq ($$(symlink),yes)
 	@mkdir -p $(prefix)/include/pony/detail
 	$(SILENT)ln $(symlink.flags) $(destdir)/bin/ponyc $(prefix)/bin/ponyc
 	$(SILENT)ln $(symlink.flags) $(destdir)/lib/libponyrt.a $(prefix)/lib/libponyrt.a
+ifeq ($(OSTYPE),linux)
+	$(SILENT)ln $(symlink.flags) $(destdir)/lib/libponyrt-pic.a $(prefix)/lib/libponyrt-pic.a
+endif
 ifneq ($(wildcard $(destdir)/lib/libponyrt.bc),)
 	$(SILENT)ln $(symlink.flags) $(destdir)/lib/libponyrt.bc $(prefix)/lib/libponyrt.bc
 endif
@@ -610,6 +616,9 @@ uninstall:
 	-$(SILENT)rm -rf $(destdir) 2>/dev/null ||:
 	-$(SILENT)rm $(prefix)/bin/ponyc 2>/dev/null ||:
 	-$(SILENT)rm $(prefix)/lib/libponyrt.a 2>/dev/null ||:
+ifeq ($(OSTYPE),linux)
+	-$(SILENT)rm $(prefix)/lib/libponyrt-pic.a 2>/dev/null ||:
+endif
 ifneq ($(wildcard $(prefix)/lib/libponyrt.bc),)
 	-$(SILENT)rm $(prefix)/lib/libponyrt.bc 2>/dev/null ||:
 endif
@@ -661,6 +670,9 @@ deploy: test
 	@mkdir -p $(package)/usr/lib/pony/$(package_version)/lib
 	$(SILENT)cp build/release/libponyc.a $(package)/usr/lib/pony/$(package_version)/lib
 	$(SILENT)cp build/release/libponyrt.a $(package)/usr/lib/pony/$(package_version)/lib
+ifeq ($(OSTYPE),linux)
+	$(SILENT)cp build/release/libponyrt-pic.a $(package)/usr/lib/pony/$(package_version)/lib
+endif
 ifneq ($(wildcard build/release/libponyrt.bc),)
 	$(SILENT)cp build/release/libponyrt.bc $(package)/usr/lib/pony/$(package_version)/lib
 endif
@@ -668,6 +680,9 @@ endif
 	$(SILENT)cp src/libponyrt/pony.h $(package)/usr/lib/pony/$(package_version)/include
 	$(SILENT)cp src/common/pony/detail/atomics.h $(package)/usr/lib/pony/$(package_version)/include/pony/detail
 	$(SILENT)ln -s /usr/lib/pony/$(package_version)/lib/libponyrt.a $(package)/usr/lib/libponyrt.a
+ifeq ($(OSTYPE),linux)
+	$(SILENT)ln -s /usr/lib/pony/$(package_version)/lib/libponyrt-pic.a $(package)/usr/lib/libponyrt-pic.a
+endif
 ifneq ($(wildcard /usr/lib/pony/$(package_version)/lib/libponyrt.bc),)
 	$(SILENT)ln -s /usr/lib/pony/$(package_version)/lib/libponyrt.bc $(package)/usr/lib/libponyrt.bc
 endif

--- a/src/libponyc/codegen/genexe.c
+++ b/src/libponyc/codegen/genexe.c
@@ -207,6 +207,8 @@ static bool link_exe(compile_t* c, ast_t* program,
   const char* ponyrt = c->opt->runtimebc ? "" :
 #if defined(PLATFORM_IS_WINDOWS)
     "ponyrt.lib";
+#elif defined(PLATFORM_IS_LINUX)
+    c->opt->pic ? "-lponyrt-pic" : "-lponyrt";
 #else
     "-lponyrt";
 #endif

